File: static/js/virtual_memory.js
// Virtual Memory Module

document.addEventListener('DOMContentLoaded', function() {
    let accessPattern = [];
    
    // Initialize with sample data
    initializePage();
    
    // Event Listeners
    document.getElementById('addAccessBtn').addEventListener('click', () => addAccess());
    document.getElementById('generateRandomBtn').addEventListener('click', generateRandomAccesses);
    document.getElementById('loadSampleBtn').addEventListener('click', loadSampleData);
    document.getElementById('simulateBtn').addEventListener('click', runSimulation);
    document.getElementById('translateBtn').addEventListener('click', translateAddress);
    document.getElementById('clearBtn').addEventListener('click', clearAll);
    
    function initializePage() {
        loadSampleData();
    }
    
    function addAccess(address = null, type = 'read') {
        const accessPatternList = document.getElementById('accessPatternList');
        const accessAddress = address !== null ? address : 0;
        const accessType = type;
        
        const accessItem = document.createElement('div');
        accessItem.className = 'access-item fade-in';
        accessItem.innerHTML = `
            <input type="number" class="access-address" placeholder="Address (decimal)" value="${accessAddress}" min="0">
            <select class="access-type">
                <option value="read" ${accessType === 'read' ? 'selected' : ''}>Read</option>
                <option value="write" ${accessType === 'write' ? 'selected' : ''}>Write</option>
            </select>
            <button class="remove-btn" onclick="removeAccess(this)">
                <i class="fas fa-times"></i>
            </button>
        `;
        
        accessPatternList.appendChild(accessItem);
        accessPattern.push({ address: accessAddress, type: accessType });
    }
    
    window.removeAccess = function(button) {
        const accessItem = button.parentElement;
        accessItem.remove();
        updateAccessPatternArray();
    };
    
    function updateAccessPatternArray() {
        accessPattern = [];
        const accessItems = document.querySelectorAll('.access-item');
        accessItems.forEach(item => {
            const address = parseInt(item.querySelector('.access-address').value);
            const type = item.querySelector('.access-type').value;
            if (!isNaN(address) && address >= 0) {
                accessPattern.push({ address, type });
            }
        });
    }
    
    function generateRandomAccesses() {
        const accessPatternList = document.getElementById('accessPatternList');
        accessPatternList.innerHTML = '';
        accessPattern = [];
        
        const virtualSize = parseInt(document.getElementById('virtualSize').value);
        const numAccesses = 10;
        
        for (let i = 0; i < numAccesses; i++) {
            const randomAddress = Math.floor(Math.random() * virtualSize);
            const randomType = Math.random() > 0.5 ? 'read' : 'write';
            addAccess(randomAddress, randomType);
        }
        
        appUtils.showNotification('Random access pattern generated!', 'success');
    }
    
    function loadSampleData() {
        const accessPatternList = document.getElementById('accessPatternList');
        accessPatternList.innerHTML = '';
        accessPattern = [];
        
        document.getElementById('virtualSize').value = 65536;
        document.getElementById('physicalSize').value = 16384;
        document.getElementById('pageSize').value = 4096;
        
        const sampleAccesses = [
            { address: 8192, type: 'read' },
            { address: 12288, type: 'write' },
            { address: 8192, type: 'read' },
            { address: 20480, type: 'read' },
            { address: 0, type: 'write' },
            { address: 8192, type: 'read' },
            { address: 4096, type: 'read' },
            { address: 20480, type: 'write' },
            { address: 12288, type: 'read' },
            { address: 16384, type: 'read' }
        ];
        
        sampleAccesses.forEach(access => addAccess(access.address, access.type));
        
        appUtils.showNotification('Sample data loaded successfully!', 'success');
    }
    
    async function runSimulation() {
        updateAccessPatternArray();
        
        const virtualSize = parseInt(document.getElementById('virtualSize').value);
        const physicalSize = parseInt(document.getElementById('physicalSize').value);
        const pageSize = parseInt(document.getElementById('pageSize').value);
        
        // Validation
        if (!appUtils.validatePositiveNumber(virtualSize, 'Virtual Memory Size')) return;
        if (!appUtils.validatePositiveNumber(physicalSize, 'Physical Memory Size')) return;
        if (!appUtils.validatePositiveNumber(pageSize, 'Page Size')) return;
        
        if (accessPattern.length === 0) {
            appUtils.showNotification('Please add at least one memory access', 'warning');
            return;
        }
        
        const data = {
            virtualSize: virtualSize,
            physicalSize: physicalSize,
            pageSize: pageSize,
            accessPattern: accessPattern
        };
        
        try {
            const result = await appUtils.makeAPIRequest('/api/virtual-memory/simulate', 'POST', data);
            displayResults(result);
            appUtils.showNotification('Simulation completed successfully!', 'success');
        } catch (error) {
            console.error('Simulation error:', error);
        }
    }
    
    function displayResults(result) {
        // Display metrics
        const metricsDisplay = document.getElementById('metricsDisplay');
        metricsDisplay.style.display = 'block';
        
        document.getElementById('pageHitsValue').textContent = result.page_hits;
        document.getElementById('pageFaultsValue').textContent = result.page_faults;
        document.getElementById('hitRateValue').textContent = result.hit_rate + '%';
        document.getElementById('diskWritesValue').textContent = result.disk_writes;
        document.getElementById('eatValue').textContent = result.effective_access_time.toFixed(2) + ' ns';
        document.getElementById('totalAccessesValue').textContent = result.total_accesses;
        
        // Display visualization
        const visualizationDisplay = document.getElementById('visualizationDisplay');
        visualizationDisplay.style.display = 'block';
        document.getElementById('chartImage').src = result.plot;
        
        // Display page table (only valid pages)
        displayPageTable(result.page_table);
        
        // Display access log
        displayAccessLog(result.access_log);
        
        appUtils.animateElement(metricsDisplay);
        appUtils.animateElement(visualizationDisplay);
    }
    
    function displayPageTable(pageTable) {
        const pageTableDisplay = document.getElementById('pageTableDisplay');
        const tableBody = document.getElementById('pageTableBody');
        
        pageTableDisplay.style.display = 'block';
        tableBody.innerHTML = '';
        
        // Only show pages that are in memory
        const validPages = pageTable.filter(page => page.valid);
        
        if (validPages.length === 0) {
            const row = document.createElement('tr');
            row.innerHTML = '<td colspan="5" style="text-align: center;">No pages in memory</td>';
            tableBody.appendChild(row);
            return;
        }
        
        validPages.forEach(page => {
            const row = document.createElement('tr');
            row.innerHTML = `
                <td><strong>${page.page_number}</strong></td>
                <td>${page.frame_number}</td>
                <td><span class="status-badge status-valid">Yes</span></td>
                <td><span class="status-badge ${page.dirty ? 'status-fault' : 'status-hit'}">${page.dirty ? 'Yes' : 'No'}</span></td>
                <td><span class="status-badge ${page.reference ? 'status-valid' : 'status-invalid'}">${page.reference ? 'Yes' : 'No'}</span></td>
            `;
            tableBody.appendChild(row);
        });
        
        appUtils.animateElement(pageTableDisplay);
    }
    
    function displayAccessLog(accessLog) {
        const accessLogDisplay = document.getElementById('accessLogDisplay');
        const tableBody = document.getElementById('accessLogBody');
        
        accessLogDisplay.style.display = 'block';
        tableBody.innerHTML = '';
        
        accessLog.forEach((log, index) => {
            const row = document.createElement('tr');
            
            const resultClass = log.page_fault ? 'status-fault' : 'status-hit';
            const resultText = log.result;
            
            row.innerHTML = `
                <td><strong>${index + 1}</strong></td>
                <td>${log.address}</td>
                <td>${log.page_number}</td>
                <td>${log.offset}</td>
                <td>${log.frame_number !== undefined ? log.frame_number : 'N/A'}</td>
                <td>${log.physical_address !== undefined ? log.physical_address : 'N/A'}</td>
                <td><span class="status-badge ${log.type === 'read' ? 'status-valid' : 'status-fault'}">${log.type.toUpperCase()}</span></td>
                <td><span class="status-badge ${resultClass}">${resultText}</span></td>
            `;
            tableBody.appendChild(row);
        });
        
        appUtils.animateElement(accessLogDisplay);
    }
    
    async function translateAddress() {
        const logicalAddress = parseInt(document.getElementById('logicalAddress').value);
        const pageSize = parseInt(document.getElementById('pageSize').value);
        
        if (isNaN(logicalAddress) || logicalAddress < 0) {
            appUtils.showNotification('Please enter a valid logical address', 'warning');
            return;
        }
        
        // For translation, we need a page table
        // We'll use the current simulation's page table if available
        // Otherwise, create a simple example
        
        const pageNumber = Math.floor(logicalAddress / pageSize);
        const offset = logicalAddress % pageSize;
        
        // Create a simple page table for demonstration
        const pageTable = {};
        // Assume first few pages are in memory
        for (let i = 0; i <= pageNumber; i++) {
            pageTable[i] = {
                valid: i < 4, // First 4 pages in memory
                frame_number: i < 4 ? i : null
            };
        }
        
        const data = {
            logicalAddress: logicalAddress,
            pageSize: pageSize,
            pageTable: pageTable
        };
        
        try {
            const result = await appUtils.makeAPIRequest('/api/virtual-memory/translate', 'POST', data);
            displayTranslationResult(result);
        } catch (error) {
            console.error('Translation error:', error);
        }
    }
    
    function displayTranslationResult(result) {
        const translationResult = document.getElementById('translationResult');
        translationResult.style.display = 'block';
        
        if (result.success) {
            translationResult.innerHTML = `
                <h4><i class="fas fa-check-circle" style="color: #4CAF50;"></i> Translation Successful</h4>
                <p><strong>Logical Address:</strong> ${result.logical_address} (0x${result.logical_address.toString(16).toUpperCase()})</p>
                <p><strong>Page Number:</strong> ${result.page_number}</p>
                <p><strong>Offset:</strong> ${result.offset}</p>
                <p><strong>Frame Number:</strong> ${result.frame_number}</p>
                <p><strong>Physical Address:</strong> ${result.physical_address} (0x${result.physical_address.toString(16).toUpperCase()})</p>
                ${result.binary_breakdown ? `
                    <hr style="margin: 1rem 0; border: none; border-top: 1px solid #ddd;">
                    <p><strong>Binary Breakdown:</strong></p>
                    <p style="font-family: monospace; font-size: 0.9rem;">
                        Logical: ${result.binary_breakdown.logical_binary}<br>
                        Page: ${result.binary_breakdown.page_bits} | Offset: ${result.binary_breakdown.offset_bits}<br>
                        Frame: ${result.binary_breakdown.frame_bits} | Offset: ${result.binary_breakdown.offset_bits}<br>
                        Physical: ${result.binary_breakdown.physical_binary}
                    </p>
                ` : ''}
            `;
        } else {
            translationResult.innerHTML = `
                <h4><i class="fas fa-exclamation-circle" style="color: #F44336;"></i> Translation Failed</h4>
                <p><strong>Logical Address:</strong> ${result.logical_address}</p>
                <p><strong>Error:</strong> ${result.error}</p>
            `;
        }
        
        appUtils.animateElement(translationResult);
    }
    
    function clearAll() {
        const accessPatternList = document.getElementById('accessPatternList');
        accessPatternList.innerHTML = '';
        accessPattern = [];
        
        document.getElementById('virtualSize').value = 65536;
        document.getElementById('physicalSize').value = 16384;
        document.getElementById('pageSize').value = 4096;
        document.getElementById('logicalAddress').value = '';
        
        document.getElementById('translationResult').style.display = 'none';
        
        appUtils.clearResults();
        appUtils.showNotification('All data cleared', 'info');
    }
});

File: modules/virtual_memory.py
import math

class VirtualMemorySimulator:
    def __init__(self):
        pass
    
    def simulate(self, virtual_size, physical_size, page_size, access_pattern):
        """
        Simulate virtual memory management
        """
        # Calculate number of virtual pages and physical frames
        num_virtual_pages = virtual_size // page_size
        num_physical_frames = physical_size // page_size
        
        # Initialize page table (all pages initially not in memory)
        page_table = {}
        for i in range(num_virtual_pages):
            page_table[i] = {
                'page_number': i,
                'frame_number': None,
                'valid': False,
                'dirty': False,
                'reference': False
            }
        
        # Track frames
        frames = [None] * num_physical_frames
        free_frames = list(range(num_physical_frames))
        
        # Simulate access pattern
        access_log = []
        page_faults = 0
        page_hits = 0
        disk_writes = 0
        
        for access in access_pattern:
            logical_address = access['address']
            access_type = access.get('type', 'read')  # 'read' or 'write'
            
            page_number = logical_address // page_size
            offset = logical_address % page_size
            
            if page_number >= num_virtual_pages:
                access_log.append({
                    'address': logical_address,
                    'page_number': page_number,
                    'offset': offset,
                    'type': access_type,
                    'result': 'Invalid address',
                    'page_fault': False
                })
                continue
            
            page_entry = page_table[page_number]
            
            if page_entry['valid']:
                # Page hit
                page_hits += 1
                frame_number = page_entry['frame_number']
                physical_address = frame_number * page_size + offset
                page_entry['reference'] = True
                
                if access_type == 'write':
                    page_entry['dirty'] = True
                
                access_log.append({
                    'address': logical_address,
                    'page_number': page_number,
                    'offset': offset,
                    'frame_number': frame_number,
                    'physical_address': physical_address,
                    'type': access_type,
                    'result': 'Hit',
                    'page_fault': False
                })
            else:
                # Page fault
                page_faults += 1
                
                if free_frames:
                    # Allocate a free frame
                    frame_number = free_frames.pop(0)
                else:
                    # Need to replace a page (using FIFO for simplicity)
                    victim_page = None
                    for pg_num, pg_entry in page_table.items():
                        if pg_entry['valid']:
                            victim_page = pg_num
                            break
                    
                    frame_number = page_table[victim_page]['frame_number']
                    
                    # Check if victim page is dirty (needs to be written back)
                    if page_table[victim_page]['dirty']:
                        disk_writes += 1
                    
                    # Invalidate victim page
                    page_table[victim_page]['valid'] = False
                    page_table[victim_page]['frame_number'] = None
                    page_table[victim_page]['dirty'] = False
                
                # Load page into frame
                frames[frame_number] = page_number
                page_entry['frame_number'] = frame_number
                page_entry['valid'] = True
                page_entry['reference'] = True
                
                if access_type == 'write':
                    page_entry['dirty'] = True
                
                physical_address = frame_number * page_size + offset
                
                access_log.append({
                    'address': logical_address,
                    'page_number': page_number,
                    'offset': offset,
                    'frame_number': frame_number,
                    'physical_address': physical_address,
                    'type': access_type,
                    'result': 'Page Fault',
                    'page_fault': True
                })
        
        # Calculate metrics
        total_accesses = len(access_pattern)
        page_fault_rate = (page_faults / total_accesses * 100) if total_accesses > 0 else 0
        hit_rate = (page_hits / total_accesses * 100) if total_accesses > 0 else 0
        
        # Effective Access Time (EAT)
        memory_access_time = 100  # nanoseconds
        page_fault_service_time = 8000000  # 8ms in nanoseconds
        eat = (hit_rate/100) * memory_access_time + (page_fault_rate/100) * page_fault_service_time
        
        return {
            'virtual_size': virtual_size,
            'physical_size': physical_size,
            'page_size': page_size,
            'num_virtual_pages': num_virtual_pages,
            'num_physical_frames': num_physical_frames,
            'page_table': [page_table[i] for i in range(num_virtual_pages)],
            'frames': frames,
            'access_log': access_log,
            'total_accesses': total_accesses,
            'page_faults': page_faults,
            'page_hits': page_hits,
            'disk_writes': disk_writes,
            'page_fault_rate': round(page_fault_rate, 2),
            'hit_rate': round(hit_rate, 2),
            'effective_access_time': round(eat, 2)
        }
    
    def translate_address(self, logical_address, page_size, page_table):
        """
        Translate a single logical address to physical address
        """
        page_number = logical_address // page_size
        offset = logical_address % page_size
        
        # Convert page_table dict keys to integers if they're strings
        page_table_int = {int(k): v for k, v in page_table.items()}
        
        if page_number in page_table_int:
            page_entry = page_table_int[page_number]
            
            if page_entry['valid']:
                frame_number = page_entry['frame_number']
                physical_address = frame_number * page_size + offset
                
                return {
                    'logical_address': logical_address,
                    'page_number': page_number,
                    'offset': offset,
                    'frame_number': frame_number,
                    'physical_address': physical_address,
                    'success': True,
                    'binary_breakdown': {
                        'logical_binary': bin(logical_address)[2:].zfill(16),
                        'page_bits': bin(page_number)[2:].zfill(8),
                        'offset_bits': bin(offset)[2:].zfill(8),
                        'frame_bits': bin(frame_number)[2:].zfill(8),
                        'physical_binary': bin(physical_address)[2:].zfill(16)
                    }
                }
            else:
                return {
                    'logical_address': logical_address,
                    'page_number': page_number,
                    'offset': offset,
                    'error': 'Page fault - Page not in memory',
                    'success': False
                }
        else:
            return {
                'logical_address': logical_address,
                'page_number': page_number,
                'offset': offset,
                'error': 'Invalid page number',
                'success': False
            }
    
    def generate_sample_access_pattern(self, num_accesses=20, max_address=65536):
        """
        Generate a sample access pattern for testing
        """
        import random
        access_pattern = []
        
        for _ in range(num_accesses):
            access_pattern.append({
                'address': random.randint(0, max_address - 1),
                'type': random.choice(['read', 'write'])
            })
        
        return access_pattern